Exits simulation mode when player wins or losses.

alkorolyov opened the pull request:
Hi,

My name is Alex and this is my first commit ever on open source. I am mainly interested in machine learning so unciv turn base game seems to be a good starting point for trying out different strategies. In order to do so, I am planning to add some features needed to create environment suitable for machine learning. For the moment I have no professional programming experience so please don't judge too harshly if something went wrong.

Back to the current feature. I used it in two player simulation when one of the players already lost and there there is no point in continuing the simulation. Can I change simulateUntilTurnForDebug mechanics or it's better to add separate var simulateUntilWinOrLose?

Also I have a foregoing plans on what could be done, like automated game simulation between AI, spectator Mode, replay recording and playing, adding units on MapEditor etc. Are there any channel/forum/chat where I can ask dumb questions before starting to work on each feature? Would be great if you could give me your view on how some features should be implemented before I actually start coding them. So any comment any additional information would be much appreciated :)

Thanks.

AcridBrimistic commented on core/src/com/unciv/logic/GameInfo.kt:
```diff
@@ -95,6 +95,17 @@ class GameInfo {
+                    println("Simulation stopped on turn $turns")
+                    for (civ in UncivGame.Current.gameInfo.civilizations) {
+                        val victoryType = civ.victoryManager.hasWonVictoryType()
+                        if (civ.victoryManager.hasWon()) {println("$civ won $victoryType victory")}
```
Superfluous curly braces.
```suggestion
if (civ.victoryManager.hasWon()) println("$civ won $victoryType victory")
```

AcridBrimistic commented on core/src/com/unciv/UncivGame.kt:
```diff
@@ -40,10 +40,11 @@ class UncivGame(
     val superchargedForDebug = false
 
     /** Simulate until this turn on the first "Next turn" button press.
+     *  Also finishes simulation when player wins or looses.
```
Typo.

```suggestion
     *  Also finishes simulation when player wins or loses.
```

AcridBrimistic commented on core/src/com/unciv/logic/GameInfo.kt:
```diff
@@ -95,6 +95,17 @@ class GameInfo {
                         && !gameParameters.noBarbarians
                         && turns % 10 == 0) placeBarbarians()
+
+                // exit simulation mode when player wins or looses
```
Typo.

```suggestion
                // exit simulation mode when player wins or loses
```

yairm210 commented:
Hi, thanks for joining us!
A. I think it's best to have this as an entirely separate parameter - the "simulate turns" is used by developers to "timeskip" to a certain era to check mechanics and units, not to check AI, and I think the overlap between the two is small enough for us to add in another parameter
B. We have a Discord server, https://discord.gg/bjrB4Xw
Be aware that 99% are players and not developers, if you're looking for developer feedback then Github is probably better =)
C. We already kind of have a spectator mode in the form of "start a new game, kill your settler, spectate", but I'd be glad to have a more "formal" spectator mode - lots of players have requested this as well ;)
D. We actually already have "place units" in the Map Editor, if you want to uncomment that - but it's not ready for public consumption quite yet (what happens if you place units of civ A but start a gave without A, etc.)

alkorolyov commented:
Hi,

Understood, so I will be asking about feature implementation here)
As for the initial commit, I am also quite new to GitHub system. Should I continue to work in this merge request or abandon it and start in new one?

I have added a separate var simulateUntilWinOrLose, to be able to simulate until game end or max turns. 

I have a general question about nextTurnAutomation. Is it possible to start automation right after game creation screen. Like not waiting for NextTurn button to be pressed. How would you implement that? Can I call GameInfo.NextTurn() right after new game creation and rendering of initial state? Where would you put this call?

I have also added alertBattle flag to be able to switch off console notification for battles. Might speed up automated battles between AI's. Is it good to add this feature in this commit or better in the separate pull request?

yairm210 commented:
You can continue this PR, I think that the required changes are not large.
As for your second question, it depends on what you're aiming for. What exactly are you attempting? A visual comparison? A metrics-based comparison? (AI 1 vs 1) compared to (AI 2 vs 2) or (AI 1 vs 2)?
If what you're aiming for is a "run until someone loses" then I think you can start activating the next turns right after the game creation at GameStarter. Or will you be loading a game from a midpoint and continuing from there?

alkorolyov commented:
For the moment I am aiming to train AI vs AI (1 vs 1) with as simple starting conditions as possible and being able run a lot of those simulations.
First, I would like to be able to run simulation from start, zero turn (AI 1 vs 1) until one of two game ending conditions:
1) someone loses or wins
2) turns exceed max turns - draw

As for the starting point I am using saved game (0 turn) to have fixed starting locations, nations and minimal map. But in the future it would be also good to have a possibility to run simulation multiple simulations from random generated maps..
Thank you for the hints, tommorow will be exploring GameStarter class and try to add NextTurn call if UncivGame.simulateUntilWinOrLose != 0. 

In the future I would like to write a function or class to be able to start simulation of two AI's without graphics and just record move order and final result - win, lose or draw.

alkorolyov commented:
Hello Yair,

I just made a few tries with adding NextTurn just after game creation. In general it worked, but not so smoothly with rendering. I believe I have messed some multithreading or smth. So for this PL I want to leave a simple automation after pressing NextTurn button and not overcomplicate things. When I will be more confident with source code I will try to add NextTurn automation directly after game creation or loading in the next PL.

For the current commit, do you think it's acceptable to merge now? 

I have also question about One city challenge. Is it supposed to apply only for human players? I am also using it in automations, but I have noticed that for the moment it applies only to human and not AI's. Should I create additional parameter to include AI in one city challenge or we can change it's general behavior?

yairm210 commented:
One city challenge applies only to the human players

If you plan on having a non-visual next turn, it's perfectly possible to create a game and run it to the end without the UI getting part of it at all - we've tried to keep the business logic separated from the UI precisely for this reason :)

alkorolyov commented:
Great, thank you for quick response.

I would like to try attack unit placing in Map editing more "public friendly". Should I create feature request for this or directly PL to discuss it there? Want to know your ideas how to improve it.

alkorolyov commented:
Sorry I forgot to set up defaults to zero for new parameter:
in UncivGame class
    var simulateUntilWinOrLose: Int = 0 
In the current commit it's 1000, which is probably not good for default option. How can I fix it?
